fppanalysis/two_dim_velocity_estimates.py

```diff
@@ -0,0 +1,140 @@
+    if hasattr(ds, "t"):
+        return np.meshgrid(ds.x.values, ds.y.values)
+    raise "Unknown format"
+
+
+def get_signal(x, y, ds):
+    # Sajidah's format
+    if hasattr(ds, "time"):
+        return ds.isel(x=x, y=y).dropna(dim="time", how="any")['frames'].values
+    # 2d code
+    if hasattr(ds, "t"):
+        return ds.isel(x=x, y=y).dropna(dim="t", how="any")['n'].values
+    raise "Unknown format"
+
+
+def get_dt(ds):
+    # Sajidah's format
+    if hasattr(ds, "time"):
+        times = ds['time']
+        return times[1].values - times[0].values
+    # 2d code
+    if hasattr(ds, "t"):
+        times = ds['t']
+        return times[1].values - times[0].values
+    raise "Unknown format"
+
+
+def estimate_velocities_for_pixel(x, y, ds):
+    """
+    Estimates radial and poloidal velocity for a pixel with indexes x,y using the pixels located
+    at (x,y), (x+1, y) and (x, y-1). Time delay estimation is performed by maximizing the cross-
+    correlation function.
+
+    Returns:
+        vx Radial velocity
+        vy Poloidal velocity
+    """
+    x_down, y_down = x, y - 1
+    x_right, y_right = x + 1, y
+    dt = get_dt(ds)
+
+    R0, Z0 = get_rz(x, y, ds)
+    R_right, Z_right = get_rz(x_right, y_right, ds)
+    R_down, Z_down = get_rz(x_down, y_down, ds)
+    frames_0 = get_signal(x, y, ds)
+    frames_right = get_signal(x_right, y_right, ds)
+    frames_down = get_signal(x_down, y_down, ds)
+    if len(frames_0) == 0 or len(frames_right) == 0 or len(frames_down) == 0:
+        return 0, 0
+
+    delta_vert = abs(Z0 - Z_down)
+    delta_tvert = tde.estimate_time_delay_ccmax(x=frames_0, y=frames_down, dt=dt)
+    if Z0 < Z_down:
+        delta_tvert = -delta_tvert
+
+    delta_hor = abs(R_right - R0)
+    delta_thor = tde.estimate_time_delay_ccmax(x=frames_right, y=frames_0, dt=dt)
+
+    vx, vy = get_2d_velocities_from_time_delays(delta_thor, delta_tvert, delta_hor, delta_vert)
+
+    return vx, vy
+
+
+def estimate_velocity_field(ds):
+    """
+    Given a dataset ds with GPI data in a format produced by https://github.com/sajidah-ahmed/cmod_functions,
+    computed the velocity field. The estimation takes into account poloidal flows as described in
+    (ask Juan). For each pixel, the velocities are estimated using the given pixel, and two neighbour
+    pixels: the right neighbour and the down neighbour. The return objects are matrices of the size of the
+    GPI grid, from which the velocity field can be easily plotted via f.e matplotlib.quiver.
+
+    Returns:
+        vx Radial velocities
+        vy Poloidal velocities
+        R Radial positions
+        Z Radial positions
+    """
+    shape = (len(ds.x.values), len(ds.y.values))
+    vx = np.zeros(shape=shape)
+    vy = np.zeros(shape=shape)
+    R, Z = get_rz_full(ds)
+
+    for i in range(0, shape[0] - 1):
+        for j in range(1, shape[1]):
+            try:
+                vx[i, j], vy[i, j] = estimate_velocities_for_pixel(i, j, ds)
+            except:
```

Comment:
One could raise a specific error (`ValueError`, `TypeError`, etc) which would make the error messager easier to understand

Reply:
Yes, but I don't know a priori what type of error is caused. I put the try/catch block to allow to compute the fields even if some pixels might be problematic for some whatever reason. If this gives an error, one should run the single pixel method to get the error stacktrace -> most likely the error will be due to something with data handling or issues with the used data. 
I will add something a bit more useful to the message